README.md
## VariantBreak - Structural variant analyzer for data visualization on VariantMap
[![Build Status](https://travis-ci.com/cytham/variantbreak.svg?branch=master)](https://travis-ci.com/cytham/variantbreak)
[![PyPI pyversions](https://img.shields.io/pypi/pyversions/variantbreak)](https://pypi.org/project/variantbreak/)
[![PyPI versions](https://img.shields.io/pypi/v/variantbreak)](https://pypi.org/project/variantbreak/)
[![Conda](https://img.shields.io/conda/v/bioconda/variantbreak)](https://anaconda.org/bioconda/variantbreak)
[![Github release](https://img.shields.io/github/v/release/cytham/variantbreak?include_prereleases)](../../releases)
[![PyPI license](https://img.shields.io/pypi/l/variantbreak)](./LICENSE.txt)

VariantBreak is a python package that integrates all structural variants (SVs) from a cohort of 
[NanoVar](https://github.com/cytham/nanovar) VCF files or variant BED files for visualization on [VariantMap](https://github.com/cytham/variantmap) or summarized into a CSV file. It also annotates and filters all SVs across all samples according to
 user input GTF/GFF/BED files. Gene annotation files can be found [here](https://www.gencodegenes.org/human/).

### Basic capabilities
* Intersects and merges all SV breakends from a sample cohort using [NanoVar](https://github.com/cytham/nanovar) VCF files 
(NanoVar-v1.3.6 or above) or variant BED files.
* Annotates each SV according to input GTF/GFF files or BED annotation files.
* Filters SVs by adding a "HIT" or "MISS" label according to input BED filter files.
* Creates a master pandas dataframe to store all data. 
* Creates a HDF5 file containing the master dataframe and some metadata which can be graphically visualized on VariantMap
 within Dash Bio.

## Getting Started

### Quick run

##### Command-line usage:
```
variantbreak [Options] -a annotation.gff3 -f filter.bed variant_path working_dir 
```

| Parameter | Argument | Comment |
| :--- | :--- | :--- |
| `-a` | annotation.gff3 | path to single annotation file or directory containing annotation files of GTF/GFF or BED formats |
| `-f` | filter.bed | path to single filter file or directory containing filter files of BED format|
| - | variant_path | path to single variant file or directory containing variant files of VCF or BED formats|
| - | working_dir | path to working directory |

##### Python console usage:
```
# Import variantbreak function from variantbreak package
from variantbreak import variantbreak

# Run variantbreak on your samples with annotation and filter files
df = variantbreak("/path/to/sample_dir/",
                  "/path/to/annotation_dir/",
                  "/path/to/filter_dir/")


# To save data to files
# Import write_to_file from variantbreak package
from variantbreak import write_to_files

# Specify dataframe variable, output file path and prefix, and delimiter of choice
write_to_files(df,
               "/path/to/output_prefix",
               sep="\t")

```
#### Output
| Output file | Comment |
| :--- | :--- |
| output.h5 | HDF5 file required for data visualization by VariantMap |
| output.csv | CSV file for data viewing, separated by the delimiter set by user |
| legend.txt | File containing the legend of the sample labels used in analysis|

For more information, see [wiki](https://github.com/cytham/variantbreak/wiki).

### Operating system: 
* Linux (x86_64 architecture, tested in Ubuntu 16.04)

### Installation:
There are three ways to install VariantBreak:
#### Option 1: Conda (Recommended)
```
# Installing from bioconda automatically installs all dependencies 
conda install -c bioconda variantbreak
```
#### Option 2: Pip (See dependencies below)
```
# Installing from PyPI requires own installation of dependencies, see below
pip install variantbreak
```
#### Option 3: GitHub (See dependencies below)
```
# Installing from GitHub requires own installation of dependencies, see below
git clone https://github.com/cytham/variantbreak.git 
cd variantbreak
pip install .
```

### Installation of dependencies
* bedtools >=2.26.0 (required to be in PATH by pybedtools)
* pybedtools >=0.8.1
* pandas >=1.0.3
* tables >=3.6.1
* fastcluster >=1.1.26

##### 1. _bedtools_
Please visit [here](https://bedtools.readthedocs.io/en/latest/content/installation.html) for instructions to install.

##### 2. _pybedtools_
Please visit [here](https://daler.github.io/pybedtools/main.html) for instructions to install.

##### 3. _pandas_
Please visit [here](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html) for instructions to install.

##### 4. _tables_
```
pip install tables
```
or
```
conda install -c conda-forge pytables
```

##### 5. _fastcluster_
```
pip install fastcluster
```
or
```
conda install -c conda-forge fastcluster
```

## Documentation
See [wiki](https://github.com/cytham/variantbreak/wiki) for more information.

## Versioning
See [CHANGELOG](./CHANGELOG.txt)

## Citation
Not available

## Author

* **Tham Cheng Yong** - [cytham](https://github.com/cytham)

## License

VariantBreak is licensed under GNU General Public License - see [LICENSE.txt](./LICENSE.txt) for details.

## Limitations
* Current version only allows input of VCF files generated by NanoVar. We will create a format adaptor in future versions to
 encompass VCF files generated by other SV callers.
 
* Processing speed of large sample cohorts has not been tested. Currently, it takes about 30 minutes to process about 100,000
 merged SVs. 


CHANGELOG.txt
VariantBreak Changelog


Release Summary:


Version 1.0.5 - Oct ?, 2021
    * Add TE option from NanoVar 1.4.0
    * Add working directory setup


Version 1.0.4 - Nov 25, 2020
    * Add badges in README.md
    * Correct typo in input.py (max_annotation)
    * Added Index column label for CSV file
    * Added Chr, Start, and End columns for SV coordinates
    * Removed spaces and extra '/' in "Annotation" of hovers


Version 1.0.3 - June 8, 2020
    * Changed default '--merged_buffer' value from 50 to 400
    * Changed gene name and gene type dividers from ',' to '/' to avoid delimiter confusing
    * Variants are now NOT sorted by prevalence, instead, it is clustered by hierarchical clustering
    * Added option '--cluster_sample' for hierarchical clustering of sample order
    * Added option '--auto_filter' for automatic removal of variants intersected with filter files instead of just labeling them


Version 1.0.2 - Apr 30, 2020
    * Changed SV classes to actual class labels in output.csv file instead of decimal representation


Version 1.0.1 - Apr 26, 2020
    * Fixed duplicated sample ID bug for BNDs in variant_parse
    * Added new dependency 'tables' for pandas hdf
    * Added ';' at the end of gene names to mark thr end of string
    * Added gene name check for ';' character
    * Added logging.critical logging for Exceptions throughout package
    * Fixed setup.py installation by changing version import approach


Version 1.0.0 - Apr 23, 2020
    * First release


variantbreak/input.py
"""
Functions for parsing and verifying input parameters.

Copyright (C) 2020 Tham Cheng Yong

This file is part of VariantBreak.

VariantBreak is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

VariantBreak is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with VariantBreak.  If not, see <https://www.gnu.org/licenses/>.
"""

import sys
import argparse

import pandas as pd
from variantbreak import __version__


# Parse input
def input_parser(args=sys.argv[1:]):
    parser = argparse.ArgumentParser(description="VariantBreak is a structural variant (SV) breakend analyzer that \
intersects and merges \nSV breakends from NanoVar VCF files or variant BED files for visualization on VariantMap \nor \
summarized into a CSV file. It also annotates and filters all SVs across all samples \naccording to user input GTF/GFF/BED \
files.",
                                     formatter_class=argparse.RawTextHelpFormatter)

    def restrict_int(f):
        try:
            if int(f) < -1:
                raise argparse.ArgumentTypeError("%r needs to be positive or -1 to disable" % (int(f),))
            else:
                return f
        except ValueError:
            raise argparse.ArgumentTypeError("%r needs to be an integer" % (f,))

    def one_character(s):
        df = pd.DataFrame()
        _sep = bytes(s, "utf-8").decode("unicode_escape")
        try:
            df.to_csv(sep=_sep)
            return _sep
        except TypeError:
            raise argparse.ArgumentTypeError("%r needs to be a single character" % (s,))

    parser.add_argument("v_file_dir", type=str,
                        metavar="[variant_path]",
                        help="""path to single variant file or directory containing variant files 
of VCF (NanoVar-v1.3.6 or above) or BED formats. Formats: .vcf/.bed""")

    parser.add_argument("wk_dir", type=str,
                        metavar="[working_directory]",
                        help="""path to working directory. Directory will be created if it does not 
exist.""")

    parser.add_argument("-a", "--annotation_file_dir", type=str,
                        metavar="path",
                        help="""path to single annotation file or directory containing annotation 
files of GTF/GFF or BED formats. Formats: .gtf/.gff/.gff3/.bed""")

    parser.add_argument("-f", "--filter_file_dir", type=str,
                        metavar="path",
                        help="""path to single filter file or directory containing filter files of 
BED format. Format: .bed""")

    parser.add_argument("-d", "--del_annotate_size", type=restrict_int, metavar="int",
                        default=20000,
                        help="""Deletions with sizes lower or equal to this value will have the 
entire deleted region annotated. Any genes that intersect with 
the deleted region will be included as annotation. On the contrary, 
if deletion size is greater than this value, only the deletion 
breakends will be annotated, omitting any annotation of the middle 
deleted region. In other words, only genes intersecting with the 
deletion breakends will be included as annotation. This is done to 
reduce false annotations due to false large deletions. Note that 
the value to be set is an absolute deletion size, do not use minus 
'-'. Use value '-1' to disable this threshold and annotate all 
deleted regions despite of size. [20000]""")

    parser.add_argument("-b", "--merge_buffer", type=int, metavar="int",
                        default=400,
                        help="""nucleotide length buffer for SV breakend clustering [400]""")

    parser.add_argument("-p", "--promoter_size", type=int, metavar="int",
                        default=1000,
                        help="""length in base-pairs upstream of TSS to define promoter region 
[1000]""")

    parser.add_argument("-m", "--max_annotation", type=int, metavar="int",
                        default=3,
                        help="""maximum number of annotation entries to be recorded in the 
dataframe for each SV [3]""")

    parser.add_argument("--cluster_sample", action='store_true',
                        help="performs hierarchical clustering on samples")

    parser.add_argument("--auto_filter", action='store_true',
                        help="""automatically removes variants that intersected
with all filter BED files""")

    parser.add_argument("-s", "--sep", type=one_character, metavar="str",
                        default=',',
                        help="""single character field delimiter for output dataframe CSV file 
(e.g. '\\t' for tab or ',' for comma) [,]""")

    parser.add_argument("-n", "--filename", type=str, metavar="str",
                        default="output",
                        help="File name prefix of output files [output]")

    parser.add_argument("-v", "--version", action='version',
                        version=__version__,
                        help="show version and exit")

    parser.add_argument("-q", "--quiet", action='store_true',
                        help="hide verbose")

    args = parser.parse_args(args)
    return args


variantbreak/valid.py
"""
Functions for validating input files.

Copyright (C) 2020 Tham Cheng Yong

This file is part of VariantBreak.

VariantBreak is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

VariantBreak is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with VariantBreak.  If not, see <https://www.gnu.org/licenses/>.
"""

import os
from distutils.version import StrictVersion
import logging


def valid(
        variant_file_dir,
        annotation_file_dir,
        filter_file_dir,
):
    """Return a dictionary with file path or name list from different input directories.

Keyword arguments:

- variant_file_dir (path; required): Variant file or path to directory containing all variant
    files of .vcf or .bed formats.
- annotation_file_dir (path): Annotation file or path to directory containing all annotation
    files of .gff, .gff3, .gtf, .bed formats.
- filter_file_dir (path): Filter file or path to directory containing all filter files of .bed format.
    """

    # Get file lists
    if os.path.isfile(variant_file_dir):
        file_dict = {'variant_list': [variant_file_dir],
                     'variant_name_list': [os.path.basename(variant_file_dir)]}
    else:
        file_dict = {'variant_list': sorted([os.path.join(variant_file_dir, file)
                                             for file in os.listdir(variant_file_dir)
                                             if os.path.isfile(os.path.join(variant_file_dir, file))]),
                     'variant_name_list': sorted([file for file in os.listdir(variant_file_dir)
                                                  if os.path.isfile(os.path.join(variant_file_dir, file))])}

    if annotation_file_dir:
        if os.path.isfile(annotation_file_dir):
            file_dict['label_list'] = [annotation_file_dir]
            file_dict['label_name_list'] = [os.path.basename(annotation_file_dir)]
        else:
            file_dict['label_list'] = sorted([os.path.join(annotation_file_dir, file)
                                              for file in os.listdir(annotation_file_dir)
                                              if os.path.isfile(os.path.join(annotation_file_dir, file))])
            file_dict['label_name_list'] = sorted([file for file in os.listdir(annotation_file_dir)
                                                   if os.path.isfile(os.path.join(annotation_file_dir, file))])
    else:
        file_dict['label_list'], file_dict['label_name_list'] = [], []

    if filter_file_dir:
        if os.path.isfile(filter_file_dir):
            file_dict['filter_list'] = [filter_file_dir]
            file_dict['filter_name_list'] = [os.path.basename(filter_file_dir)]
        else:
            file_dict['filter_list'] = sorted([os.path.join(filter_file_dir, file)
                                               for file in os.listdir(filter_file_dir)
                                               if os.path.isfile(os.path.join(filter_file_dir, file))])
            file_dict['filter_name_list'] = sorted([file for file in os.listdir(filter_file_dir)
                                                    if os.path.isfile(os.path.join(filter_file_dir, file))])
    else:
        file_dict['filter_list'], file_dict['filter_name_list'] = [], []

    # Validate files
    for file in file_dict['variant_list']:
        if file.endswith('.vcf'):
            valid_vcf(file)
        elif file.endswith('.bed'):
            valid_bed(file)
        else:
            logging.critical('Error: Variant directory has file format(s) other than .vcf and .bed, such as %s' % file)
            raise Exception('Error: Variant directory has file format(s) other than .vcf and .bed, such as %s' % file)
    label_gtf = 0
    for file in file_dict['label_list']:
        if file.endswith('.gff') or file.endswith('.gff3') or file.endswith('.gtf'):
            label_gtf += 1
            if label_gtf > 1:
                logging.critical('Error: Only a single input gtf/gff file is allow.')
                raise Exception('Error: Only a single input gtf/gff file is allow.')
            # valid_gff(file)
        elif file.endswith('.bed'):
            valid_bed(file)
        else:
            logging.critical('Error: Annotation directory has file format(s) other than .gff, .gtf and .bed, such as %s' % file)
            raise Exception('Error: Annotation directory has file format(s) other than .gff, .gtf and .bed, such as %s' % file)
    for file in file_dict['filter_list']:
        if file.endswith('.bed'):
            valid_bed(file)
        else:
            logging.critical('Error: Filter directory has file format(s) other than .bed, such as %s' % file)
            raise Exception('Error: Filter directory has file format(s) other than .bed, such as %s' % file)
    return file_dict


def valid_vcf(file):
    with open(file) as f:
        for line in f:
            first = line
            _ = next(f)
            third = next(f)
            vcf_version = float(first.split('v')[1])
            if vcf_version != 4.2:  # fileformat=VCFv4.2
                logging.critical('Error: VCF file %s version is not 4.2' % file)
                raise Exception('Error: VCF file %s version is not 4.2' % file)
            nv_version = third.split('-')[1]
            if StrictVersion('1.3.6') > StrictVersion(nv_version):  # source=NanoVar-1.3.6
                logging.critical('Error: VCF file %s was generated by an outdated NanoVar version, version needs to be >= 1.3.6' %
                                 file)
                raise Exception('Error: VCF file %s was generated by an outdated NanoVar version, version needs to be >= 1.3.6' %
                                file)
            break


def valid_bed(file):
    n = 0
    with open(file) as f:
        for line in f:
            n += 1
            if not line.split('\t')[0].startswith('chr'):
                logging.critical('Error: Contig names has to start with "chr" but detected %s, line %s of %s' % (line.split(
                    '\t')[0], str(n), file))
                raise Exception('Error: Contig names has to start with "chr" but detected %s, line %s of %s' % (line.split(
                    '\t')[0], str(n), file))
            try:
                start = int(line.split('\t')[1])
                end = int(line.split('\t')[2])
            except TypeError:
                logging.critical('Error: Non-integer value at line %s of %s' % (str(n), file))
                raise Exception('Error: Non-integer value at line %s of %s' % (str(n), file))
            if start > end:
                logging.critical('Error: End value greater than Start value at line %s of %s' % (str(n), file))
                raise Exception('Error: End value greater than Start value at line %s of %s' % (str(n), file))


# def valid_gff(file):
#     return None


scripts/filter_variants_by_prevalence.py
import os
import pandas as pd
import math
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
from matplotlib.ticker import ScalarFormatter
import scipy.stats as stats
import fastcluster

filename = "data.csv"
df0 = pd.read_csv(filename, delimiter='\t', index_col = 0)
df = df0[~((df0['Filter3'] == 1) & (df0['End'] - df0['Start'] < 100))]  # Only remove SV in homopolyer if SV < 100 bp

sample_names = ['S1', 'S2', 'S3', 'S4', 'S5', 'S6', 'S7', 'S8', 'S9', 'S10', 'S11', 'S12', 'S13', 'S14', 'S15', 'S16',
                'S17', 'S18', 'S19', 'S20', 'S21', 'S22', 'S23', 'S24']

depth = [
    7.55,
    8.14,
    7.66,
    3.68,
    9.38,
    3.42,
    4.9,
    7.18,
    4.38,
    4.44,
    4.23,
    9.36,
    11.18,
    9.85,
    8.09,
    6.61,
    9.88,
    9.1,
    9.04,
    7.0,
    8.25,
    11.64,
    8.06,
    7.81
]

depth_dict = {}
for i,j in zip(sample_names, depth):
    depth_dict[i] = j

# Create list of only samples with depth > 6
sample_list = []
for i in sample_names:
    if depth_dict[i] > 6:
        sample_list.append(i)

# Create list of non-AML sample names
normal = ['S18', 'S19', 'S20', 'S21', 'S22', 'S23', 'S24']

# Create lists of AML sample names
aml = []
for name in sample_list:
    if name in normal:
        pass
    else:
        aml.append(name)

# Get non_AML array from dataframe
non_aml_arr = df.loc[:, normal].values

# Convert all SV hits to 1 or misses to 0
non_aml_arr[non_aml_arr != '-'] = 1
non_aml_arr[non_aml_arr == '-'] = 0

# Count SV hits and misses
non_aml_out = []
for i in non_aml_arr:
    non_aml_out.append([np.count_nonzero(i == 0), np.count_nonzero(i == 1)])

# Get AML array from dataframe
aml_arr = df.loc[:, aml].values

# Convert all SV hits to 1 or misses to 0
aml_arr[aml_arr != '-'] = 1
aml_arr[aml_arr == '-'] = 0

# Count SV hits and misses
aml_out = []
for i in aml_arr:
    aml_out.append([np.count_nonzero(i == 0), np.count_nonzero(i == 1)])

# Combine AML and non-AML and tranpose list, generating list of 2D arrays
out = []
for i in zip(aml_out, non_aml_out):
    out.append(list(map(list, zip(*i))))

total_aml = len(aml)
total_norm = len(normal)



# A) Detect AML-associated variants by prevalence [CHOOSE A OR B]
aml_thres = 0.8  # Change alpha values to adjust stringency
norm_thres = 0.2  # Change beta values to adjust stringency
a = 0
indexes = []
memory = []
for i in out:
    if i[1][0]/total_aml >= aml_thres and i[1][1]/total_norm <= norm_thres:
        indexes.append(a)
        memory.append([a, i[1][0]/total_aml, i[1][1]/total_norm])
    a += 1


# B) Detect normal-associated variants by prevalence [CHOOSE A OR B]
aml_thres = 0.2  # Change beta values to adjust stringency
norm_thres = 0.8 # Change alpha values to adjust stringency
a = 0
indexes_norm = []
for i in out:
    if i[1][0]/total_aml <= aml_thres and i[1][1]/total_norm >= norm_thres:
        indexes_norm.append(a)
        memory.append([a, i[1][0] / total_aml, i[1][1] / total_norm])
    a += 1



select_df = df.iloc[indexes + indexes_norm, :]

# Cluster using linkage hclustering
# Create numpy array for selected samples
arr = select_df.loc[:, aml + normal].values

# Change all SV hits to value of 1 and misses to 0
arr[arr != '-'] = 1
arr[arr == '-'] = 0

# Change astype to int
arr = arr.astype(int)
z = fastcluster.linkage(arr, 'ward')
n = len(z) + 1
cache = dict()
for k in range(len(z)):
    c1, c2 = int(z[k][0]), int(z[k][1])
    c1 = [c1] if c1 < n else cache.pop(c1)
    c2 = [c2] if c2 < n else cache.pop(c2)
    cache[n + k] = c1 + c2

index = cache[2 * len(z)]
sv_order = select_df.index.values.tolist()
indexes1 = []
for i in index:
    indexes1.append(sv_order[i])

# Get default metadata
with pd.HDFStore('data.h5', mode="r") as store:
    df_default = store['dataset']
    metadata_default = store.get_storer('dataset').attrs.metadata

new_df = df_default.reindex(index=indexes1)

# Save dataframe to CSV
new_df.to_csv('new.csv', index=True,
              sep='\t',
              index_label='Index')

# Save dataframe and metadata to HDF5 for viewing on VariantMap
store = pd.HDFStore('new.h5', mode='w')
store.put('dataset', new_df)
store.get_storer('dataset').attrs.metadata = metadata_default
store.close()




variantbreak/variant_break.py
"""
Main annotation function of VariantBreak.

Copyright (C) 2020 Tham Cheng Yong

This file is part of VariantBreak.

VariantBreak is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

VariantBreak is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with VariantBreak.  If not, see <https://www.gnu.org/licenses/>.
"""

from __future__ import absolute_import

import os
import sys
import logging
from datetime import datetime

import csv
import pandas as pd
from .valid import valid
from .parse import variant_parse, annotation_parse, filter_parse


def variantbreak(
        variant_file_dir,
        annotation_file_dir=None,
        filter_file_dir=None,
        # ignore_duplicate=False,
        del_annotate_size=20000,
        merge_buffer=50,
        promoter_size=1000,
        max_annotation=3,
        cluster_sample=False,
        auto_filter=False,
        quiet=False
):
    """ Returns a dataframe with metadata

- variant_file_dir (str; required): Path to a single variant file or directory containing
    variant files in VCF or BED formats. VCF files needs to be generated by NanoVar long-read
    SV caller (v1.3.5 or above). BED files need to be tab-delimited with the essential three
    columns: chrom, chromStart and chromEnd (column 4 will be ignored).
- annotation_file_dir (str; optional): Path to a single annotation file or directory containing
    annotation files of GTF/GFF or BED formats.
- filter_file_dir (str; optional): Path to a single filter file or directory containing filter
    files of BED format.
- del_annotate_size (int; default 20000): Deletions with sizes lower or equal to this value will
    have the entire deleted region annotated. Any genes that intersect with the deleted region
    will be included as annotation. On the contrary, if deletion size is greater than this value,
    only the deletion breakends will be annotated, omitting any annotation of the middle deleted
    region. In other words, only genes intersecting with the deletion breakends will be included
    as annotation. This is to reduce false annotations due to false large deletions. Note that
    the value to be set is an absolute deletion size, do not use minus '-'. Use value '-1' to
    disable this threshold and annotate all deleted regions despite of size.
- merge_buffer (int; default 50): Nucleotide length buffer for SV breakend clustering
- promoter_size (int; default 1000): Length in base-pairs upstream of TSS to define promoter region
- max_annotation (int; default 3): Maximum number of annotation entries to be recorded in the
    dataframe for each SV
- cluster_sample (bool; default False): If True, performs hierarchical clustering on samples.
- auto_filter (bool; default False): If True, automatically removes variants that intersected
    with all filter BED files.
- quiet (bool; default False): Hide verbose if True.


Usage example:

# Import variantbreak function from variantbreak package
from variantbreak import variantbreak

# Run variantbreak on your samples in sample_dir, together with annotation and filter files, and output dataframe to df
df = variantbreak("/path/to/sample_dir/",
                  "/path/to/annotation_dir/",
                  "/path/to/filter_dir/")

# To save data to files
# Import write_to_file from variantbreak package
from variantbreak import write_to_files

# Specify dataframe variable, output file path and prefix, and delimiter
write_to_files(df,
               "/path/to/output_prefix",
               sep="\t")

# Output files
output.h5 - HDF5 file required for data visualization by VariantMap
output.csv - CSV file for data viewing, separated by the delimiter set by user
legend.txt - File containing the legend of the sample labels used in the dataframe

    """

    # Set verbose
    if quiet:
        sys.stdout = open(os.devnull, 'w')

    # Disable del_annotate_size if it is -1
    if int(del_annotate_size) == -1:
        del_annotate_size = float('inf')

    # Check file integrity and right formatting for input files
    file_dict = valid(variant_file_dir, annotation_file_dir, filter_file_dir)

    # Print and record time started
    now = datetime.now()
    now_str = now.strftime("[%d/%m/%Y %H:%M:%S]")
    print(now_str, "- VariantBreak started")
    logging.info('VariantBreak started')

    # Parse annotation files
    annotation_dict, gene_info_dict, label_col, annote_name_dict = annotation_parse(file_dict, promoter_size)

    # Parse filter files
    filter_dict, filter_col, filter_name_dict = filter_parse(file_dict)

    # Parse and intersect variant files and generate master dataframe
    df, sample_name_dict = variant_parse(file_dict,
                                         del_annotate_size,
                                         merge_buffer,
                                         annotation_dict,
                                         gene_info_dict,
                                         label_col,
                                         filter_dict,
                                         filter_col,
                                         max_annotation,
                                         cluster_sample,
                                         auto_filter
                                         )

    # Gather sample labels
    sample_names = [x for x in sample_name_dict.keys()]

    # Add metadata to dataframe
    df.metadata = ''  # To avoid pandas UserWarning
    df.metadata = {'sample_names': sample_names, 'sample': sample_name_dict, 'annotation': annote_name_dict,
                   'filter': filter_name_dict}

    # Print and record time finished
    now = datetime.now()
    now_str = now.strftime("[%d/%m/%Y %H:%M:%S]")
    print(now_str, "- VariantBreak ended")
    logging.info('VariantBreak ended')
    return df


def write_to_files(df, file_path, sep=','):

    # Save dataframe and metadata to HDF5
    filename = file_path  # , _ = os.path.splitext(file_path)
    store = pd.HDFStore(filename + '.h5', mode='w')
    store.put('dataset', df)
    metadata = df.metadata
    store.get_storer('dataset').attrs.metadata = metadata
    store.close()

    # Save dataframe to CSV
    df2 = df.copy()
    for name in metadata['sample_names']:
        df2[name].replace({0.0: "-", 0.214: "DEL", 0.357: "INV", 0.500: "INS", 0.643: "BND", 0.786: "DUP", 0.929: "UKN"},
                          inplace=True)
    df2.to_csv(filename + '.csv', index=True, sep=sep, index_label='Index')

    # Save label names to file_path in a legend text file for variant and filter files
    w = open(os.path.join(os.path.dirname(filename), 'legend.txt'), 'w')
    w.write('#Label' + sep + 'File_path\n')
    write_name = csv.writer(w, delimiter=sep)
    legend = df.metadata['sample']
    legend.update(df.metadata['filter'])
    legend.update(df.metadata['annotation'])
    for key, val in legend.items():
        write_name.writerow([key, val])
    w.close()
